_posts/2019-07-31-neuron2.md

```diff
@@ -0,0 +1,276 @@
+---
+title:  "Training our Neuron. Algorithms to keep it on track."
+date: 2019-07-31 00:00:00 +0800
+author: jimmy
+---
+
+Part 2 of my foray into machine learning. In part one we built a neuron class in python. In this post we'll start to 
+train it to be useful... or as useful as a binary classifier can be.
+
+<!--more-->
+
+If you haven't read [Part 1]({% post_url 2019-07-16-neuron %}) I would highly suggest you start there. As a quick recap
+though, here is what we accomplished by the end of the last post.
+
+We built a neuron class that looks like the following:
+```python
+import random
+
+class Neuron:
+    def __init__(self, number_of_inputs, weights=None):
+        self.number_of_inputs = number_of_inputs
+        self.weights = weights if weights is not None else self._random_weights()
+
+    def _random_weights(self):
+        return [random.normalvariate(0, .5) for _ in range(self.number_of_inputs)]
+
+    def _weighted_sum(self, inputs):
+        return sum([inputs[i] * self.weights[i] for i in range(self.number_of_inputs)])
+
+    def _activate(self, w_sum):
+        return -1 if w_sum < 0 else 1
+
+    def predict(self, inputs):
+        w_sum = self._weighted_sum(inputs)
+        return self._activate(w_sum)
+```
+
+We can instantiate it, and then get it to make predictions from inputs, but those predictions are basically a coin toss,
+based on the random weights assigned at initialization.
+
+```python
+n = Neuron(number_of_inputs=2)
+n.predict([100, 50])
+```
+Running this code a dozen times should yield a pretty random split of 1s and 0s. The goal of this post is to reduce
+that randomness it steer it closer and closer to a correct prediction.
+
+# Forward the Foundation
+
+### Baseline
+To start, I created a bunch of random data for my perceptron to work with. Without doing any training, I ran it through
+several attempts at classifying the data. Here are a few of its outcomes:
+<div>
+    <img src="/assets/images/articles/initial_performance.png">
```

Comment:
I've been following this pattern for images:

```
<div style="text-align: center;">
    <img src="/assets/images/articles/sprint-process.png" class="img-bordered">
</div>
```

Gives them a lil border, centers if not full width, etc. -- may not be super useful in this case, kind of annoying I have to look this snippet up each time as well..